cadc-access-control-server/src/main/java/ca/nrc/cadc/ac/server/oidc/OIDCUtil.java

```diff
@@ -0,0 +1,202 @@
+ *  but WITHOUT ANY WARRANTY;            sera utile, mais SANS AUCUNE
+ *  without even the implied             GARANTIE : sans même la garantie
+ *  warranty of MERCHANTABILITY          implicite de COMMERCIALISABILITÉ
+ *  or FITNESS FOR A PARTICULAR          ni d’ADÉQUATION À UN OBJECTIF
+ *  PURPOSE.  See the GNU Affero         PARTICULIER. Consultez la Licence
+ *  General Public License for           Générale Publique GNU Affero
+ *  more details.                        pour plus de détails.
+ *
+ *  You should have received             Vous devriez avoir reçu une
+ *  a copy of the GNU Affero             copie de la Licence Générale
+ *  General Public License along         Publique GNU Affero avec
+ *  with OpenCADC.  If not, see          OpenCADC ; si ce n’est
+ *  <http://www.gnu.org/licenses/>.      pas le cas, consultez :
+ *                                       <http://www.gnu.org/licenses/>.
+ *
+ ************************************************************************
+ */
+package ca.nrc.cadc.ac.server.oidc;
+
+import ca.nrc.cadc.ac.Group;
+import ca.nrc.cadc.ac.Role;
+import ca.nrc.cadc.ac.User;
+import ca.nrc.cadc.ac.UserNotFoundException;
+import ca.nrc.cadc.ac.server.GroupPersistence;
+import ca.nrc.cadc.ac.server.UserPersistence;
+import ca.nrc.cadc.ac.server.ldap.LdapGroupPersistence;
+import ca.nrc.cadc.ac.server.ldap.LdapUserPersistence;
+import ca.nrc.cadc.auth.AuthenticationUtil;
+import ca.nrc.cadc.auth.HttpPrincipal;
+import ca.nrc.cadc.auth.NumericPrincipal;
+import ca.nrc.cadc.auth.SignedToken;
+import ca.nrc.cadc.net.TransientException;
+
+import java.io.IOException;
+import java.net.URI;
+import java.security.AccessControlException;
+import java.security.InvalidKeyException;
+import java.security.Key;
+import java.security.KeyPair;
+import java.util.ArrayList;
+import java.util.Calendar;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+
+import javax.security.auth.Subject;
+
+import org.apache.log4j.Logger;
+
+import io.jsonwebtoken.JwtBuilder;
+import io.jsonwebtoken.Jwts;
+import io.jsonwebtoken.SignatureAlgorithm;
+import io.jsonwebtoken.security.Keys;
+
+/**
+ * 
+ * Utilities for OIDC/OAuth2 support.
+ * 
+ * @author majorb
+ *
+ */
+public class OIDCUtil {
+    
+    public static final String AUTHORIZE_TOKEN_SCOPE = "cadc:oauth2/authorize_token";
+    public static final String REFRESH_TOKEN_SCOPE = "cadc:oauth2/refresh_token";
+    public static final String ACCESS_TOKEN_SCOPE = "cadc:oauth2/access_token";
+    
+    public static final Integer ID_TOKEN_EXPIRY_MINUTES = 60*24*7*2; // 2 weeks
+    public static final Integer AUTHORIZE_CODE_EXPIRY_MINUTES = 10;
+    public static final Integer ACCESS_CODE_EXPIRY_MINUTES = 60*24*7*2; // 2 weeks
+    public static final Integer REFRESH_TOKEN_EXPIRY_MINUTES = 60*24*7*52; // 1 year
+    public static final Integer JWT_EXPIRY_MINUTES = 60*24*7*2; // 2 weeks
+    
```

Comment:
Hard coded URL in constant below. Should that value be determined by finding a resourceID in LocalAuthority, looking up a resourceID, then looking up the service? If you want the base URL you could lookup the capabilities URL and then strip off the last path element (since that has to be there for this to work anyway).

Reply:
Ah, yes, missed that.

Reply:
I took this approach but used the 'oauth' capability to find the base URL instead.